workers/gossip-server/src/peer_discovery/heartbeat.rs

```diff
@@ -115,45 +110,106 @@ impl GossipProtocolExecutor {
     // | Helpers |
     // -----------
 
+    // --- Heartbeat --- //
+
+    /// Records a successful heartbeat
+    pub(super) async fn record_heartbeat(
+        &self,
+        peer_id: &WrappedPeerId,
+    ) -> Result<(), GossipError> {
+        Ok(self.state.record_heartbeat(peer_id).await?)
+    }
+
+    /// Build a heartbeat message
+    pub async fn build_heartbeat(&self) -> Result<HeartbeatMessage, GossipError> {
+        Ok(self.state.construct_heartbeat().await?)
+    }
+
+    // --- Peer Expiry --- //
+
+    /// Update the expiry status of a peer
+    async fn update_expiry_status(&self, peer_id: WrappedPeerId) -> Result<(), GossipError> {
+        if self.expiry_buffer.is_expiry_candidate(&peer_id).await {
+            self.maybe_expire_candidate(peer_id).await
+        } else {
+            self.maybe_expire_peer(peer_id).await
+        }
```

Comment:
IIUC, it looks like the expiration of a cluster peer now goes through two stages via `update_expiry_status` - if the peer hasn't heartbeated sufficiently recently, it gets marked as an expiry candidate, then if it is a candidate it will be expired on the next call to `update_expiry_status`. `update_expiry_status` is only called in `send_heartbeat`, so we basically may wait an extra heartbeat round, leaving a peer in a candidate state, before we actually expire it. This seems fine to me, just wanted to be sure I'm understanding correctly

Reply:
I suppose this is the reason why we add peers to the invisibility window when we mark them as candidates - to allow for the extra heartbeat round?